System/FilePath/Glob/Directory.hs

```diff
@@ -321,3 +321,16 @@ commonDirectory = second unseparate . splitP . separate
    fromConst d (ExtSeparator   :xs) = fromConst (d `DL.snoc` extSeparator) xs
    fromConst d (LongLiteral _ s:xs) = fromConst (d `DL.append`DL.fromList s) xs
    fromConst _ _                    = Nothing
+
+   -- "." must be explicitly matched at the start of a pattern or after a path
+   -- separator; this function checks for an implicit match i.e. "[.]".
+   startsWithImplicitExt p =
+      case unPattern p of
+         (Literal '.' : _) ->
+            True
+         (LongLiteral _ l : _) ->
+            case l of
+               ('.':_) -> True
+               _ -> False
```

Comment:
This could be simplified to `LongLiteral _ "." : _ -> True`, but I'd simply leave it out as `optimize` should guarantee that there are no length-1 `LongLiteral`s.

Reply:
I tried that to start with, but I don't think that will always work; consider:

```haskell
> compile "[.]foo" == Pattern [LongLiteral 4 ".foo"]
True
```

Reply:
In fact, I should add that as a test case.

Reply:
Ah, true, my mistake.

It might be worth dedicating a new `Token` constructor for `[.]` to make things simpler... but let's just go with this for now.